ViewEnvironmentUI/Sources/ViewEnvironmentPropagating.swift

```diff
@@ -0,0 +1,144 @@
+    /// bridging from other propagation systems like WorkflowUI).
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var environmentAncestor: ViewEnvironmentPropagating? { get }
+
+    /// The [`ViewEnvironment` propagation](x-source-tag://ViewEnvironmentObserving)
+    /// descendants.
+    ///
+    /// This describes the descendants that will be notified when the `ViewEnvironment` changes.
+    ///
+    /// To override the return value of this property for `UIViewController`/`UIView` subclasses, set the
+    /// ``ViewEnvironmentPropagatingObject/environmentDescendantsOverride`` property.  If no override is present, the
+    /// return value will be a collection of all `children` in  addition to the `presentedViewController` for
+    /// `UIViewController`s and `subviews` for `UIView`s.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var environmentDescendants: [ViewEnvironmentPropagating] { get }
+}
+
+extension ViewEnvironmentPropagating {
+    /// The `ViewEnvironment` that is flowing through the propagation hierarchy.
+    ///
+    /// If you'd like to provide overrides for the environment as it flows through a node, you should conform to
+    /// `ViewEnvironmentObserving` and provide those overrides in `customize(environment:)`. E.g.:
+    /// ```swift
+    /// func customize(environment: inout ViewEnvironment) {
+    ///     environment.traits.mode = .dark
+    /// }
+    /// ```
+    ///
+    /// By default, this property gets the environment by recursively walking to the root of the
+    /// propagation path, and applying customizations on the way back down. The invalidation path may be
+    /// interrupted if a node has set it's `environmentAncestor` to `nil`, even if there is a node
+    /// which specifies this node as an `environmentDescendant`.
+    ///
+    /// If you'd like to update the return value of this variable and have those changes propagated through the
+    /// propagation hierarchy, conform to `ViewEnvironmentObserving` and call ``setNeedsEnvironmentUpdate()`` and wait
+    /// for the system to call `apply(context:)` when appropriate (e.g. on the next layout pass for
+    /// `UIViewController`/`UIView` subclasses).
+    ///
+    /// - Important: `UIViewController` and `UIView` conformers _must_ call
+    /// ``ViewEnvironmentObserving/applyEnvironmentIfNeeded()-8gr5k`` in `viewWillLayoutSubviews()` and
+    /// `layoutSubviews()` respectively.
+    ///
+    public var environment: ViewEnvironment {
+        var environment = environmentAncestor?.environment ?? .empty
+
+        if let observing = self as? ViewEnvironmentObserving {
+            observing.customize(environment: &environment)
+        }
+
+        return environment
+    }
+
+    /// Notifies all appropriate descendants that the environment needs update.
+    ///
+    /// If a descendant's ancestor is `nil` it will not be notified of needing update.
```

Comment:
```suggestion
    /// Ancestor-descendent bindings must be mutually agreed for this method to notify them. If a descendant's ancestor is `nil` it will not be notified of needing update.
```

Reply:
7fdbe94